src/hooks/useMutex.js

```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+
+function useMutex() {
+  const queue = React.useRef([]);
+  const lock = React.useRef(false);
+
+  const wait = React.useCallback(() => {
+    const next = queue.current.shift();
+    if (!next) {
+      return;
+    }
+    lock.current = true;
+    return Promise.resolve(next());
+  }, []);
+
+  const signal = React.useCallback(() => {
+    lock.current = false;
+    return wait();
+  }, [wait]);
+
+  const synchronize = React.useCallback(
+    runnable => {
+      queue.current.push(runnable);
+      if (!lock.current) {
+        return wait();
+      }
+      return Promise.resolve();
+    },
+    [wait, lock]
+  );
+
+  return { synchronize, signal, wait };
+}
+
+const AnimationMutexContext = React.createContext();
+
+function AnimationProvider(props) {
+  const mutex = useMutex();
```

Comment:
This provider's value is a way for different components to access the same animation mutex.